crates/context/config/src/client/env/proxy/query/proposals.rs

```diff
@@ -31,10 +34,74 @@ impl Method<Starknet> for ProposalsRequest {
     type Returns = Vec<Proposal>;
 
     fn encode(self) -> eyre::Result<Vec<u8>> {
-        todo!()
+        let mut bytes = Vec::with_capacity(64); // 2 * 32 bytes for two u32 parameters
+
+        // First parameter (offset): pad to 32 bytes
+        bytes.extend_from_slice(&[0; 28]); // 28 bytes of zeros
+        bytes.extend_from_slice(&(self.offset as u32).to_be_bytes()); // 4 bytes of actual value
+
+        // Second parameter (length): pad to 32 bytes
+        bytes.extend_from_slice(&[0; 28]); // 28 bytes of zeros
+        bytes.extend_from_slice(&(self.length as u32).to_be_bytes()); // 4 bytes of actual value
+
+        Ok(bytes)
     }
 
-    fn decode(_response: Vec<u8>) -> eyre::Result<Self::Returns> {
-        todo!()
+    fn decode(response: Vec<u8>) -> eyre::Result<Self::Returns> {
+        // First, convert bytes back to Felts
+        let mut felts = Vec::new();
+        for chunk in response.chunks(32) {
+            if chunk.len() == 32 {
+                felts.push(Felt::from_bytes_be(chunk.try_into().unwrap()));
+            }
+        }
+        // First felt is array length
+        let array_len = felts[0].to_bytes_be()[31] as usize;
+        // println!("Array length: {}", array_len);
+
+        let mut proposals = Vec::with_capacity(array_len);
+        let mut offset = 1; // Skip the length felt
+
+        for i in 0..array_len {
+            // Each proposal starts with:
+            // - proposal_id: 2 felts (high, low)
+            // - author_id: 2 felts (high, low)
+            // - action variant: 1 felt
+            let variant = felts[offset + 4].to_bytes_be()[31];
+            // println!("Proposal {} at offset {}, variant {}", i, offset, variant);
+
+            let proposal_end = match variant {
+                0 => {
+                    // ExternalFunctionCall
+                    let calldata_len = felts[offset + 7].to_bytes_be()[31] as usize;
+                    offset + 8 + calldata_len // base + contract + selector + len + calldata
+                }
+                1 => {
+                    // Transfer
+                    offset + 8 // base + amount(2) + receiver
+                }
+                2 => {
+                    // SetNumApprovals
+                    offset + 6 // base + num
+                }
+                3 => {
+                    // SetActiveProposalsLimit
+                    offset + 6 // base + limit
+                }
+                4 => {
+                    // SetContextValue
+                    let key_len = felts[offset + 5].to_bytes_be()[31] as usize;
+                    let value_len = felts[offset + 6 + key_len].to_bytes_be()[31] as usize;
+                    offset + 7 + key_len + value_len // base + key_len + value_len + key + value
+                }
+                _ => return Err(eyre::eyre!("Unknown action variant: {}", variant)),
+            };
+
+            let proposal = StarknetProposal::decode(&felts[offset..proposal_end])
```

Comment:
does the `Decode` impl of `StarknetProposal` require the iterator to be fully exhausted?

if it doesn't why do we need the hardcoded logic for the proposal_end?

if I understand correctly and the issue is about progressively advancing through the input felts, then make an iterator ahead of time (outside the consuming loop) and pass it to `decode` on every iteration

Reply:
updated